### ISSUE TITLE
Slowdown when network packet > 10240 bytes

### ISSUE THREAD

**Issue by arneto** [siddontang/ledisdb#135]
Hi Siddontang,

Our application uses hiredis C-client to ledisdb with rocksdb.

I am investigating an issue with unexpected slow performance in our software.
I can reproduce with a simple "SET %b %b" command. I am using random 16byte keys, and play with the length of the value to test for a certain network packet length.

It boils down to the fact that if I send a network packet up to 10240 bytes (FYI: keylen=16, valuelen=10194 => packet length=10240), performance is excellent:

write socket(fd=3) 10240 bytes, 0,000029s
read socket(fd=3) 5 bytes, 0,000131s
write socket(fd=3) 10240 bytes, 0,000016s
read socket(fd=3) 5 bytes, 0,000074s
write socket(fd=3) 10240 bytes, 0,000109s
read socket(fd=3) 5 bytes, 0,000002s
write socket(fd=3) 10240 bytes, 0,000015s
read socket(fd=3) 5 bytes, 0,000056s
write socket(fd=3) 10240 bytes, 0,000015s
read socket(fd=3) 5 bytes, 0,000055s
write socket(fd=3) 10240 bytes, 0,000015s
read socket(fd=3) 5 bytes, 0,000051s
write socket(fd=3) 10240 bytes, 0,000015s
read socket(fd=3) 5 bytes, 0,000046s
write socket(fd=3) 10240 bytes, 0,000022s
read socket(fd=3) 5 bytes, 0,000044s
write socket(fd=3) 10240 bytes, 0,000015s
read socket(fd=3) 5 bytes, 0,000050s
write socket(fd=3) 10240 bytes, 0,000014s
read socket(fd=3) 5 bytes, 0,000051s
0,001386s
(and for the global time most of the time is spent printing the output)
In our live application this scenario goes well for multi-gigabytes of traffic.

However when I increment the size of the value 1 more, performance becomes bad:

write socket(fd=3) 10241 bytes, 0,000026s
read socket(fd=3) 5 bytes, 0,000153s
write socket(fd=3) 10241 bytes, 0,000014s
read socket(fd=3) 5 bytes, 0,000123s
write socket(fd=3) 10241 bytes, 0,000014s
read socket(fd=3) 5 bytes, 0,000106s
write socket(fd=3) 10241 bytes, 0,000019s
read socket(fd=3) 5 bytes, 0,000081s
write socket(fd=3) 10241 bytes, 0,000003s
read socket(fd=3) 5 bytes, 0,204138s
write socket(fd=3) 10241 bytes, 0,000009s
read socket(fd=3) 5 bytes, 0,204024s
write socket(fd=3) 10241 bytes, 0,000009s
read socket(fd=3) 5 bytes, 0,204180s
write socket(fd=3) 10241 bytes, 0,000018s
read socket(fd=3) 5 bytes, 0,203568s
write socket(fd=3) 10241 bytes, 0,000013s
read socket(fd=3) 5 bytes, 0,203813s
write socket(fd=3) 10241 bytes, 0,000007s
read socket(fd=3) 5 bytes, 0,204257s
1,225415s

The first 4 reads are still OK, but then it goes wrong - and stays bad as long as you push the server.

To outrule a network issue, I can rerun my test scenario on redisdb (default install, I guess that is just a in-memory db) and tried 10x the length without any issue like this.

Does this ring a bell somewhere?

Thanks and best regards,
Arne

(Arne Tossens)

**Comment by siddontang** [comment]
Hi @arneto 

The default read connection buffer is 10k in the config, here 

```
conn_read_buffer_size = 10240
conn_write_buffer_size = 10240
```

maybe this causes your problem, but I am confused too. 

I write a benchmark tool using go, you can use it to compare your c tool, like this:

```
ledis-benchmark -t=set -vsize=10194 -n=1 -r=10 -c=1
ledis-benchmark -t=set -vsize=10195 -n=1 -r=10 -c=1
```

the value size is10194 or 10195 ,round is 10, and request number is 1 in every round, one client.

or like this:

```
ledis-benchmark -t=set -vsize=10194 -n=1000 -c=1
ledis-benchmark -t=set -vsize=10195 -n=1000 -c=1
```

In my computer, the different value size does not affect the result too much, so I wait yours. 

Thank you.

**Comment by arneto** [comment]
Hi Siddontang,

This absolutely fixes my issue.
Your help is greatly appreciated.

Thanks and best regards,
Arne

**Comment by siddontang** [comment]
Hi @arneto 

Do you fix your problem now? 
How about the comparison between your c tool and ledis-benchmark?
Increasing the conn buffer is still slow?

Thank you! :smile:

**Comment by arneto** [comment]
Hi Siddontang,

I changed in config conn_read_buffer_size = 102400.
This is more than needed, however the number of connections is limited and under our control.

Tested on real world sizes, and responses come in really fast - also in the real application.
I will tune config conn_read_buffer_size to what is needed later.

I have not run the ledis-benchmark.

Thanks for the advice and best regards,
Arne
